src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,498 @@
+        }
+
+        public bool TryGetValue(TKey key, out TValue value)
+        {
+            if (key == null) ThrowHelper.ThrowKeyArgumenNullException();
+            Entry[] entries = _entries;
+            int collisionCount = 0;
+            for (int i = _buckets[(int)((uint)key.GetHashCode() % (uint)_buckets.Length)] - 1;
+                    (uint)i < (uint)entries.Length; i = entries[i].next)
+            {
+                if (key.Equals(entries[i].key))
+                {
+                    value = entries[i].value;
+                    return true;
+                }
+                if (collisionCount >= entries.Length)
+                {
+                    // The chain of entries forms a loop; which means a concurrent update has happened.
+                    // Break out of the loop and throw, rather than looping forever.
+                    ThrowHelper.ThrowInvalidOperationException_ConcurrentOperationsNotSupported();
+                }
+                collisionCount++;
+            }
+
+            value = default(TValue);
+            return false;
+        }
+
+        public bool Remove(TKey key)
+        {
+            if (key == null) ThrowHelper.ThrowKeyArgumenNullException();
+            Entry[] entries = _entries;
+            int bucketIndex = (int)((uint)key.GetHashCode() % (uint)_buckets.Length);
+            int entryIndex = _buckets[bucketIndex] - 1;
+
+            int lastIndex = -1;
+            while (entryIndex != -1)
+            {
+                Entry candidate = entries[entryIndex];
+                if (candidate.key.Equals(key))
+                {
+                    if (lastIndex != -1)
+                    {   // Fixup preceding element in chain to point to next (if any)
+                        entries[lastIndex].next = candidate.next;
+                    }
+                    else
+                    {   // Fixup bucket to new head (if any)
+                        _buckets[bucketIndex] = candidate.next + 1;
+                    }
+
+                    entries[entryIndex] = default; // could use RuntimeHelpers.IsReferenceOrContainsReferences
+
+                    entries[entryIndex].next = -3 - _freeList; // New head of free list
+                    _freeList = entryIndex;
+
+                    _count--;
+                    return true;
+                }
+                lastIndex = entryIndex;
+                entryIndex = candidate.next;
+            }
+
+            return false;
+        }
+
+        public ref TValue this[TKey key]
+        {
+            get
+            {
+                if (key == null) ThrowHelper.ThrowKeyArgumenNullException();
+                Entry[] entries = _entries;
+                int collisionCount = 0;
+                int bucketIndex = (int)((uint)key.GetHashCode() % (uint)_buckets.Length);
+                for (int i = _buckets[bucketIndex] - 1;
+                        (uint)i < (uint)entries.Length; i = entries[i].next)
```

Comment:
Are either of these worth trying/measuring? 
1. Extract `entries.Length` into a local outside the loop
1. Initialize `collisionCount` to `entries.Length` and decrement instead of increment, so the test is against zero rather than a value - is that cheaper?

More a matter of curiosity than expecting measurable benefits.

Reply:
@mikedn @AndyAyersMS

Reply:
Actually `(uint)i < (uint)entries.Length` could be `i >= 0` couldn't it? Buckets will never have anything other than valid indexes and 0 (meaning -1 index)

Reply:
It's a bounds check perf trick

Reply:
Its also validating `i`; otherwise `entries[i].next` will emit a check anyway

Reply:
> It's a bounds check perf trick

Unfortunately this trick doesn't work with loop conditions. Doesn't hurt either so you can leave it in, maybe I'll get to fix it one day.

> Initialize collisionCount to entries.Length and decrement instead of increment, so the test is against zero rather than a value - is that cheaper?

Unlikely to make a material difference.

> Extract entries.Length into a local outside the loop

That's almost always a bad idea.